tokio/src/sync/watch.rs

```diff
@@ -437,6 +437,46 @@ impl<T> Sender<T> {
         Ok(())
     }
 
+    /// Modifies watched value, notifying all receivers.
+    ///
+    /// This can useful for modifying the watched value, without
+    /// having to allocate a new instance. Additionally, this
+    /// method permits sending values even when there are no receivers.
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio::sync::watch;
+    ///
+    /// struct State {
+    ///     counter: usize,
+    /// }
+    /// let (state_tx, state_rx) = watch::channel(State { counter: 0 });
+    /// state_tx.send_modify(|state| state.counter += 1);
+    /// assert_eq!(state_rx.borrow().counter, 1);
+    /// ```
+    pub fn send_modify<F>(&self, func: F)
+    where
+        F: FnOnce(&mut T),
+    {
+        {
+            // Acquire the write lock and update the value.
+            let mut lock = self.shared.value.write().unwrap();
+            // Update the value.
+            func(&mut lock);
```

Comment:
One question I have about the current design is that this will poison the lock if a panic happens in `func`. We should decide on what the behavior should be in case of a panic and not poison the lock.

Reply:
I think we should attempt to catch the panic with `catch_unwind` and `send_modify` should return a Result<(), Error> in case calling `func` resulted in panic.

Reply:
Actually, I'd rather not return the panic to the user. Even if you want to catch it, we should [resume](https://doc.rust-lang.org/std/panic/fn.resume_unwind.html) it.

Reply:
That is the first I've had to catch panics so I'm completely not sure if I've implemented this correctly.